Add TA troubleshooting page (#4708)
open-telemetry/opentelemetry.io: content/en/docs/kubernetes/operator/troubleshooting/target-allocator.md

---
title: Target Allocator
cSpell:ignore: bleh targetallocator
---

If you’ve enabled
[Target Allocator](/docs/kubernetes/operator/target-allocator/) service
discovery on the [OpenTelemetry Operator](/docs/kubernetes/operator), and your
metrics aren’t getting scraped, there are a few troubleshooting steps that you
can take to help you understand what’s going on and restore normal operation.

## Troubleshooting steps

### Did you deploy all of your resources to Kubernetes?

As a first step, make sure that you have deployed all relevant resources to your
Kubernetes cluster.

### Do you know if metrics are actually being scraped?

After you’ve deployed all of your resources to Kubernetes, make sure that the

> Target Allocator is scraping metrics from your

Reviewer: ```suggestion
Target Allocator is discovering scrape targets from your
```

  },
  "serviceMonitor/opentelemetry/sm-example/0": {
    "_link": "/jobs/serviceMonitor%2Fopentelemetry%2Fsm-example%2F0/targets"
  },
  "serviceMonitor/opentelemetry/sm-example/2": {
    "_link": "/jobs/serviceMonitor%2Fopentelemetry%2Fsm-example%2F2/targets"
  },
  "podMonitor/opentelemetry/pm-example/0": {
    "_link": "/jobs/podMonitor%2Fopentelemetry%2Fpm-example%2F0/targets"
  }
}
```

Where `serviceMonitor/opentelemetry/sm-example/0` represents one of the
`Service` ports that the `ServiceMonitor`picked up:

- `opentelemetry` is the namespace in which the `ServiceMonitor` resource
  resides.
- `sm-example` is the name of the `ServiceMonitor`.
- `0` is one of the port endpoints matched between the `ServiceMonitor` and the
  `Service`.

Similarly, the `PodMonitor`, shows up as `podMonitor/opentelemetry/pm-example/0`
in the `curl` output.

This is good news, because it tells us that the service discovery is working!

You might also be wondering about the `otel-collector` entry. The Target
Allocator automatically scrapes metrics from the OTel Collector as well.

We can take a deeper look into `serviceMonitor/opentelemetry/sm-example/0`, to

> see what metrics are getting picked up by running `curl` against the value of

Reviewer: ```suggestion
see what scrape targets are getting picked up by running `curl` against the value of
```

you must explicitly enable it:

- Set `spec.targetAllocator.enabled` to `true`
- Set `spec.targetAllocator.prometheusCR.enabled` to `true`

So that your `OpenTelemetryCollector` resource looks like this:

```yaml
apiVersion: opentelemetry.io/v1beta1
kind: OpenTelemetryCollector
metadata:
  name: otelcol
  namespace: opentelemetry
spec:
  mode: statefulset
  targetAllocator:
    enabled: true
    serviceAccount: opentelemetry-targetallocator-sa
    prometheusCR:
      enabled: true
```

See the full `OpenTelemetryCollector`
[resource definition in "Do you know if metrics are actually being scraped?"](#do-you-know-if-metrics-are-actually-beingscraped).

### Did you configure a ServiceMonitor (or PodMonitor) selector?

If you configured a
[`ServiceMonitor`](https://observability.thomasriley.co.uk/prometheus/configuring-prometheus/using-service-monitors/#:~:text=The%20ServiceMonitor%20is%20used%20to,build%20the%20required%20Prometheus%20configuration.)
selector, it means that the Target Allocator only looks for `ServiceMonitors`
having a `metadata.label` that matches the value in

> [`serviceMonitorSelector`](https://github.com/open-telemetry/opentelemetry-operator/blob/main/docs/api.md#opentelemetrycollectorspectargetallocatorprometheuscr).

Author: ```suggestion
[`serviceMonitorSelector`](https://github.com/open-telemetry/opentelemetry-operator/blob/main/docs/api.md#opentelemetrycollectorspectargetallocatorprometheuscr-1).
```

Suppose that you configured a

> [`serviceMonitorSelector`](https://github.com/open-telemetry/opentelemetry-operator/blob/main/docs/api.md#opentelemetrycollectorspectargetallocatorprometheuscr)

Author: ```suggestion
[`serviceMonitorSelector`](https://github.com/open-telemetry/opentelemetry-operator/blob/main/docs/api.md#opentelemetrycollectorspectargetallocatorprometheuscr-1)
```

for your Target Allocator, like in the following example:

```yaml
apiVersion: opentelemetry.io/v1beta1
kind: OpenTelemetryCollector
metadata:
  name: otelcol
  namespace: opentelemetry
spec:
  mode: statefulset
  targetAllocator:
    enabled: true
    serviceAccount: opentelemetry-targetallocator-sa
    prometheusCR:
      enabled: true

>       serviceMonitorSelector:
>         app: my-app

Reviewer: This would be correct in apiVersion `v1alpha1`, but in `v1beta1` this field is a label selector, and should look like this:

```suggestion
      serviceMonitorSelector:
        matchLabels:
          app: my-app
```

You also have the documentation for v1alpha1 linked above, which is most likely the source of the confusion here. The v1beta1 struct is documented [here](https://github.com/open-telemetry/opentelemetry-operator/blob/main/docs/api.md#opentelemetrycollectorspectargetallocatorprometheuscr-1).

Author: Oooh, thanks for that catch!

      serviceMonitorSelector:
        app: my-app
```

It means that your `ServiceMonitor` resource must in turn have that same value

> in `metadata.labels`:

Author: Done!

  enabled: true
  podMonitorSelector: {}
  serviceMonitorSelector: {}
```

See the
[full OpenTelemetryCollector definition in "Do you know if metrics are actually being scraped?"](#do-you-know-if-metrics-are-actually-beingscraped).

### Do your labels, namespaces, and ports match for your ServiceMonitor and your Service (or PodMonitor and your Pod)?

The `ServiceMonitor` is configured to pick up Kubernetes
[Services](https://kubernetes.io/docs/concepts/services-networking/service/)
that match on:

- Labels
- Namespaces (optional)
- Ports (endpoints)

Suppose that you have this `ServiceMonitor`:

```yaml
apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
metadata:
  name: sm-example
  labels:
    app: my-app
    release: prometheus
spec:
  selector:
    matchLabels:
      app: my-app
  namespaceSelector:
    matchNames:
      - opentelemetry
  endpoints:
    - port: prom
      path: /metrics
    - port: py-client-port
      interval: 15s
    - port: py-server-port
```

The previous `ServiceMonitor` is looking for any services that have:

- the label `app: my-app`
- reside in a namespace called `opentelemetry`
- a port named `prom`, `py-client-port`, _or_ `py-server-port`

> So for example, the `Service` resource below would get picked up by the
> `ServiceMonitor`, because it matches the above criteria:

Reviewer 2: ```suggestion
For example, the following `Service` resource would get picked up by the
`ServiceMonitor`, because it matches the previous criteria:
```

```yaml
apiVersion: v1
kind: Service
metadata:
  name: py-prometheus-app
  namespace: opentelemetry
  labels:
    app: my-app
    app.kubernetes.io/name: py-prometheus-app
spec:
  selector:
    app: my-app
    app.kubernetes.io/name: py-prometheus-app
  ports:
    - name: prom
      port: 8080
```

Conversely, the following `Service` resource would NOT, because the
`ServiceMonitor` is looking for ports named `prom`, `py-client-port`, _or_
`py-server-port`, and this service’s port is called `bleh`.

```yaml
apiVersion: v1
kind: Service
metadata:
  name: py-prometheus-app
  namespace: opentelemetry
  labels:
    app: my-app
    app.kubernetes.io/name: py-prometheus-app
spec:
  selector:
    app: my-app
    app.kubernetes.io/name: py-prometheus-app
  ports:

>     - name: bleh

Reviewer 2: Some humor in docs — thanks! :D

      port: 8080
```

{{% alert title="Tip" %}}

If you’re using `PodMonitor`, the same applies, except that it picks up

> Kubernetes _pods_ that match on labels, namespaces, and named ports.

Reviewer 2: ```suggestion
Kubernetes pods that match on labels, namespaces, and named ports.
```

---
title: Troubleshooting the Target Allocator

> description: Tips for troubleshooting the Target Allocator

Reviewer 3: ```suggestion
```

The description should either provide some additional details to the title or skip it

cSpell:ignore: bleh targetallocator
---

If you’ve enabled
[Target Allocator](/docs/kubernetes/operator/target-allocator/) service
discovery on the [OpenTelemetry Operator](/docs/kubernetes/operator), and your
metrics aren’t getting scraped, then there are a few troubleshooting steps that
you can take to help you understand what’s going on and to get things back on
track.

## Troubleshooting Steps

> ### 1- Did you deploy all of your resources to Kubernetes?

Reviewer 3: ```suggestion
### Did you deploy all of your resources to Kubernetes?
```

Although this may seem obvious and straightforward, it can happen! First things
first: make sure that you have deployed all relevant resources to your
Kubernetes cluster.

> ### 2- Do you know if metrics are actually being scraped?

Reviewer 3: ```suggestion
### Do you know if metrics are actually being scraped?
```

After you’ve deployed all of your resources to Kubernetes, check to make sure
that the Target Allocator is actually scraping metrics from your

> [`ServiceMonitor`](https://observability.thomasriley.co.uk/prometheus/configuring-prometheus/using-service-monitors/#:~:text=The%20ServiceMonitor%20is%20used%20to,build%20the%20required%20Prometheus%20configuration.)(s)

Reviewer 3: Something seems off with this link. Also is there a target to an official prometheus documentation?

Author: Done!

and/or
[`PodMonitor`](https://prometheus-operator.dev/docs/user-guides/getting-started/#using-podmonitors)(s).

Suppose that you have this `ServiceMonitor` definition:

```yaml
apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
metadata:
  name: sm-example
  namespace: opentelemetry
  labels:
    app.kubernetes.io/name: py-prometheus-app
    release: prometheus
spec:
  selector:
    matchLabels:
      app: my-app
  namespaceSelector:
    matchNames:
      - opentelemetry
  endpoints:
    - port: prom
      path: /metrics
    - port: py-client-port
      interval: 15s
    - port: py-server-port
```

this `Service` definition:

```yaml
apiVersion: v1
kind: Service
metadata:
  name: py-prometheus-app
  namespace: opentelemetry
  labels:
    app: my-app
    app.kubernetes.io/name: py-prometheus-app
spec:
  selector:
    app: my-app
    app.kubernetes.io/name: py-prometheus-app
  ports:
    - name: prom
      port: 8080
```

and this `OpenTelemetryCollector` definition:

```yaml
apiVersion: opentelemetry.io/v1beta1
kind: OpenTelemetryCollector
metadata:
  name: otelcol
  namespace: opentelemetry
spec:
  mode: statefulset

>   image: otel/opentelemetry-collector-contrib:0.102.1

Reviewer 3: We might consider auto-updating this collector version here, but this can happen in a later update

Author: @svrnm I can also leave it out, and then the OTel Operator will just use whatever the default image is

Reviewer 3: if that works, I would prefer that, except someone from @open-telemetry/operator-approvers disagrees

Author: I'll hold off on updating until we get feedback from @open-telemetry/operator-approvers, then. 🙂

Reviewer: It's better to leave it blank in examples, as a rule. We do run the risk of our examples being broken by future releases, but in the case of K8s API Versioning, breaking changes can only happen with an explicit version change, the worst that can happen is that the whole resource will be rejected as unsupported. There won't be subtle breakage.

Author: Noted, and removed!

          "__meta_kubernetes_endpointslice_labelpresent_endpointslice_kubernetes_io_managed_by": "true",
          "__meta_kubernetes_service_annotation_kubectl_kubernetes_io_last_applied_configuration": "{\"apiVersion\":\"v1\",\"kind\":\"Service\",\"metadata\":{\"annotations\":{},\"labels\":{\"app\":\"my-app\",\"app.kubernetes.io/name\":\"py-otel-client\"},\"name\":\"py-otel-client-svc\",\"namespace\":\"opentelemetry\"},\"spec\":{\"ports\":[{\"name\":\"py-client-port\",\"port\":8082,\"protocol\":\"TCP\",\"targetPort\":\"py-client-port\"}],\"selector\":{\"app.kubernetes.io/name\":\"py-otel-client\"}}}\n",
          "__meta_kubernetes_pod_labelpresent_app_kubernetes_io_name": "true",
          "__meta_kubernetes_pod_container_port_protocol": "TCP",
          "__meta_kubernetes_service_label_app": "my-app",
          "__meta_kubernetes_endpointslice_labelpresent_app_kubernetes_io_name": "true"
        }
      }
    ]
  }
}
```

{{% alert title="Note" %}}

See the
[Target Allocator readme](https://github.com/open-telemetry/opentelemetry-operator/blob/b98d5a497713253082e399c4561ac539cc91c5a9/cmd/otel-allocator/README.md?plain=1#L128-L134)

> for more information on the `/jobs` endpoint.

Reviewer 3: Can we move that to docs as well? Also can we point to HEAD of main instead of a specific commit?

Author: @svrnm I've updated the link to point to main instead of the specific commit. Just to confirm - this links to the Operator's Target Allocator README. Would we want to move all contents from this readme to the docs and make that the single source of truth?

{{% /alert %}}

> ### 3- Is the Target Allocator enabled? Is Prometheus service discovery enabled?

Reviewer 3: ```suggestion
### Is the Target Allocator enabled? Is Prometheus service discovery enabled?
```

If the `curl` commands above don’t show a list of expected `ServiceMonitor`s and
`PodMonitor`s, then it’s time to dig a bit deeper.

One thing to remember is that just because you include the `targetAllocator`
section in the `OpenTelemetryCollector` CR doesn’t mean that it’s enabled. You
need to explicitly enable it. Furthermore, if you want to use
[Prometheus service discovery](https://github.com/open-telemetry/opentelemetry-operator/blob/main/cmd/otel-allocator/README.md#discovery-of-prometheus-custom-resources),
you must explicitly enable it:

- Set `spec.targetAllocator.enabled` to `true`
- Set `spec.targetAllocator.prometheusCR.enabled` to `true`

So that your `OpenTelemetryCollector` resource looks like this:

```yaml
apiVersion: opentelemetry.io/v1beta1
kind: OpenTelemetryCollector
metadata:
  name: otelcol
  namespace: opentelemetry
spec:
  mode: statefulset
  targetAllocator:
    enabled: true
    serviceAccount: opentelemetry-targetallocator-sa
    prometheusCR:
      enabled: true
```

For more detail, see the full `OpenTelemetryCollector`
[resource definition in step 2 above](#2--do-you-know-if-metrics-are-actually-beingscraped).

> ### 4- Did you configure a ServiceMonitor (or PodMonitor) selector?

Reviewer 3: ```suggestion
### Did you configure a ServiceMonitor (or PodMonitor) selector?
```

in `metadata.labels`:

```yaml
apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
metadata:
  name: sm-example
  labels:
    app: my-app
    release: prometheus
spec:
```

For more detail, see the full `ServiceMonitor`
[resource definition in step 2](#2--do-you-know-if-metrics-are-actually-beingscraped).

In this case, the `OpenTelemetryCollector` resource's
`prometheusCR.serviceMonitorSelector` is looking only for `ServiceMonitors`
having the label `app: my-app`, which we see in the previous example.

If your ServiceMonitor resource is missing that label, then the Target Allocator
won’t pick up that ServiceMonitor.

{{% alert title="Tip" %}}

The same applies if you’re using a
[PodMonitor](https://prometheus-operator.dev/docs/user-guides/getting-started/#using-podmonitors).
In that case, if your `OpenTelemetryCollector`resource defined a
[`podMonitorSelector`](https://github.com/open-telemetry/opentelemetry-operator/blob/main/docs/api.md#opentelemetrycollectorspectargetallocatorprometheuscr),
then any `PodMonitors` you wish to be picked up by the Target Allocator would
need to have that same label.

{{% /alert %}}

> ### 5- Did you leave out the serviceMonitorSelector and/or podMonitorSelector configuration altogether?

Reviewer 3: ```suggestion
### Did you leave out the serviceMonitorSelector and/or podMonitorSelector configuration altogether?
```

As we saw in

> [step 4](#4--did-you-configure-a-servicemonitor-or-podmonitor-selector), setting

Reviewer 3: ```suggestion
["Did you configure a ServiceMonitor or PodMonitor selector"](#4--did-you-configure-a-servicemonitor-or-podmonitor-selector), setting
```

mismatched values for `serviceMonitorSelector` and `podMonitorSelector` results
in your `ServiceMonitors` and `PodMonitors`, respectively, not getting picked
up.

Similarly, in
[`v1beta1`](https://github.com/open-telemetry/opentelemetry-operator/blob/main/docs/api.md#opentelemetrycollector-1)
of the `OpenTelemetryCollector` CR, leaving out this configuration altogether,
your `PodMonitors` and `ServiceMonitors` might also not get picked up.

As of `v1beta1` of the `OpenTelemetryOperator`, a `serviceMonitorSelector`, and
`podMonitorSelector` must be included, even if you don’t intend to use it, like
this:

```yaml
---
prometheusCR:
  enabled: true
  podMonitorSelector: {}
  serviceMonitorSelector: {}
```

See the
[full OpenTelemetryCollector definition in step 2](#2--do-you-know-if-metrics-are-actually-beingscraped).

> ### 6- Do your labels, namespaces, and ports match for your ServiceMonitor and your Service (or PodMonitor and your Pod)?

Reviewer 3: ```suggestion
### Do your labels, namespaces, and ports match for your ServiceMonitor and your Service (or PodMonitor and your Pod)?
```

  config:
    receivers:
      otlp:
        protocols:
          grpc: {}
          http: {}
      prometheus:
        config:
          scrape_configs:
            - job_name: 'otel-collector'
              scrape_interval: 10s
              static_configs:
                - targets: ['0.0.0.0:8888']

    processors:
      batch: {}

    exporters:
      logging:
        verbosity: detailed

    service:
      pipelines:
        traces:
          receivers: [otlp]
          processors: [batch]
          exporters: [logging]
        metrics:
          receivers: [otlp, prometheus]
          processors: []
          exporters: [logging]
        logs:
          receivers: [otlp]
          processors: [batch]
          exporters: [logging]
```

First, set up a `port-forward` in Kubernetes, so that you can expose the Target
Allocator service:

```shell
kubectl port-forward svc/otelcol-targetallocator -n opentelemetry 8080:80
```

Where `otelcol-targetallocator` is the value of `metadata.name` in your
`OpenTelemetryCollector` CR concatenated with the `-targetallocator` suffix, and
`opentelemetry` is the namespace to which the `OpenTelemetryCollector` CR is

> deployed.

Reviewer 4: we could also give people the query for the service by doing `kubectl get svc -l app.kubernetes.io/component=opentelemetry-targetallocator`

Author: Thanks! Added that.

```shell
curl localhost:8080/jobs | jq
```

Your sample output should look like this:

```json
{
  "serviceMonitor/opentelemetry/sm-example/1": {
    "_link": "/jobs/serviceMonitor%2Fopentelemetry%2Fsm-example%2F1/targets"
  },
  "otel-collector": {
    "_link": "/jobs/otel-collector/targets"
  },
  "serviceMonitor/opentelemetry/sm-example/0": {
    "_link": "/jobs/serviceMonitor%2Fopentelemetry%2Fsm-example%2F0/targets"
  },
  "serviceMonitor/opentelemetry/sm-example/2": {
    "_link": "/jobs/serviceMonitor%2Fopentelemetry%2Fsm-example%2F2/targets"
  },
  "podMonitor/opentelemetry/pm-example/0": {
    "_link": "/jobs/podMonitor%2Fopentelemetry%2Fpm-example%2F0/targets"
  }
}
```

Where `serviceMonitor/opentelemetry/sm-example/0` represents one of the
`Service` ports that the `ServiceMonitor`picked up:

- `opentelemetry` is the namespace in which the `ServiceMonitor` resource
  resides.
- `sm-example` is the name of the `ServiceMonitor`.
- `0` is one of the port endpoints matched between the `ServiceMonitor` and the
  `Service`.

Similarly, the `PodMonitor`, shows up as `podMonitor/opentelemetry/pm-example/0`
in the `curl` output.

> This is good news, because it tells us that the service discovery is working!

Reviewer 4: ```suggestion
This is good news, because it tells us that the scrape config discovery is working!
```

Service discovery working is the next step where we check that targets are showing up for these scrape_configs' service discovery.

You might also be wondering about the `otel-collector` entry. The Target

> Allocator automatically scrapes metrics from the OTel Collector as well.

Reviewer 4: this entry is because [the config provided](https://github.com/open-telemetry/opentelemetry.io/pull/4708/files#diff-1dbe631d16dc9eb3b4e3af1e41c754015d43209646e535c073e124e5c4a7e2e9R100-R103) has the otel-collector self scrape set, the TA wouldn't have it if not for that block

Author: Thanks for clarifying! Added.

We can take a deeper look into `serviceMonitor/opentelemetry/sm-example/0`, to
see what scrape targets are getting picked up by running `curl` against the
value of the `_link` output above:

```shell

> curl localhost:8080/jobs/serviceMonitor%2Fdefault%2Fmy-app%2F0/targets | jq

Reviewer 4: this is great to see – i love showing the full curl and sample output for what this should look like! one thing... this curl's URL isn't found in the previous output... I think it should be `/jobs/serviceMonitor%2Fopentelemetry%2Fsm-example%2F0/targets`

Author: Good catch! Fixed.

```

Sample output:

```json
{
  "otelcol-collector-0": {

>     "_link": "/jobs/serviceMonitor%2Fopentelemetry%2Fsm-example%2F1/targets?collector_id=otelcol-collector-0",

Reviewer 4: would be good to note here that the query parameter collector_id is saying that these are the targets for otelcol-collector-0

Author: Noted!

In that case, you would use a
[`podMonitorSelector`](https://github.com/open-telemetry/opentelemetry-operator/blob/main/docs/api.md#opentelemetrycollectorspectargetallocatorprometheuscr)
instead of a `serviceMonitorSelector`.

{{% /alert %}}

### Did you leave out the serviceMonitorSelector and/or podMonitorSelector configuration altogether?

As mentioned in
["Did you configure a ServiceMonitor or PodMonitor selector"](#did-you-configure-a-servicemonitor-or-podmonitor-selector),
setting mismatched values for `serviceMonitorSelector` and `podMonitorSelector`
results in the Target Allocator failing to discover scrape targets from your
`ServiceMonitors` and `PodMonitors`, respectively.

Similarly, in
[`v1beta1`](https://github.com/open-telemetry/opentelemetry-operator/blob/main/docs/api.md#opentelemetrycollector-1)
of the `OpenTelemetryCollector` CR, leaving out this configuration altogether
also results in the Target Allocator failing to discover scrape targets from
your `ServiceMonitors` and `PodMonitors`.

As of `v1beta1` of the `OpenTelemetryOperator`, a `serviceMonitorSelector` and
`podMonitorSelector` must be included, even if you don’t intend to use it, like

> this:

Reviewer 4: i would include that setting it like this means it will match everything.

Author: Done!

```yaml
apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
metadata:
  name: sm-example
  labels:
    app: my-app
    release: prometheus
spec:
  selector:
    matchLabels:
      app: my-app
  namespaceSelector:
    matchNames:
      - opentelemetry
  endpoints:
    - port: prom
      path: /metrics
    - port: py-client-port
      interval: 15s
    - port: py-server-port
```

The previous `ServiceMonitor` is looking for any services that have:

- the label `app: my-app`
- reside in a namespace called `opentelemetry`
- a port named `prom`, `py-client-port`, _or_ `py-server-port`

For example, the following `Service` resource would get picked up by the
`ServiceMonitor`, because it matches the previous criteria:

```yaml
apiVersion: v1
kind: Service
metadata:
  name: py-prometheus-app
  namespace: opentelemetry
  labels:
    app: my-app
    app.kubernetes.io/name: py-prometheus-app
spec:
  selector:
    app: my-app
    app.kubernetes.io/name: py-prometheus-app
  ports:
    - name: prom
      port: 8080
```

The following `Service` resource would not be picked up, because the
`ServiceMonitor` is looking for ports named `prom`, `py-client-port`, _or_

> `py-server-port`, and this service’s port is called `bleh`.

Reviewer 4: This is such a great example and _exactly_ what I helped a user debug in slack a few weeks ago! Thanks for including it.

Author: ❤️
